[AIRFLOW-6005] Common classes should always be imported from airflow, Depends on [AIRFLOW-6128]

potiuk opened the pull request:
NOTE FOR REVIEWER: This change depends on #6596 #6702 #6696  so please review only the last commit.

For the sake of consistency, all the common components exposed directly
in 'airflow' __init__.py should be imported from there rather than
from sub-packages. This is important also for back-porting
purposes (AIP-21) to make sure that we are using stable imports.

This change does it and implements pre-commit that prevents importing
those common imports differently.


Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6005

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

potiuk committed:
[AIRFLOW-6005] Common imports should be imported directly from 'airflow'

For the sake of consistency, all the common components exposed directly
in 'airflow' __init__.py should be imported from there rather than
from sub-packages. This is important also for back-porting
purposes (AIP-21) to make sure that we are using stable imports.

This change does it and implements pre-commit that prevents importing
those common imports differently.

ashb commented on airflow/deprecations/airflow_exception_deprecated.py:
```diff
@@ -0,0 +1,34 @@
+class AirflowException(AirflowExceptionsDeprecated):
+    """Deprecated version of AirflowException"""
+    def __init__(self, *args):
+        warnings.warn(
```
This is going to warn at throw time, only right? What is the pattern we used for AIP-21?

And as with you rother PR, for small classes I would just put this inside airflow/deprecations/__init__.py (or airflow/deprecations.py) -- each extra file we have to import adds a small penalty to Airflow startup time (which is already noticeable.)

potiuk replied:
Yep. that was an experiment - I tried to make docs build. And tried different approaches but originally I had It in the _init__ and I am going to put it back.

ashb commented on airflow/__init__.py:
```diff
@@ -29,15 +29,20 @@
 from airflow.models import DAG
 
+if 'BUILDING_AIRFLOW_DOCS' in os.environ:
+    from airflow.exceptions import AirflowException
```
Why do we need this when building docs?

potiuk replied:
The problem is explained above. when I had 'deprecated' AirflowEccwption I had errro about 'duplicate candidates' for autoapi. So I had to exclude the whole 'deprecations' folder from autoapi - but this caused a problem of 'missing import module' when generating docs. I tried to mock it using 'autodoc_mock_imports' but it did not work for airflow internal modules. So I found that I can use the same mechanism as used in 'apply_defaults' with the variable.

potiuk replied:
Indeed getting __all__ solved the problem. Thanks @ashb

potiuk committed:
[AIRFLOW-6128] Simplify plugins_manager and webserver plugin code

While adding type annotations, it was apparent that the code for
plugin manager and webserver initialisation was rather convoluted
and complex. This change fixes it as follow up after untangling
cyclic dependnencies.

As a side effect - it turned out that the test plugin view
in airflow RBAC implementation was not working since 1.10.1 - it threw "render
is not an attribute" error - this has been fixed in this commit
as well. The Plugins/TestView now works as expected. See
https://stackoverflow.com/questions/53701030/airflow-plugins-rbac-enabled-blueprint-not-working?answertab=active#tab-top

feluelle commented on .pre-commit-config.yaml:
```diff
@@ -237,6 +237,43 @@ repos:
+      - id: common-imports
+        language: pygrep
+        name: Make sure AirflowConfigException is imported consistently 'from airflow'
+        entry: "from airflow\\.exception import.* AirflowConfigException"
```
```suggestion
        entry: "from airflow\\.exceptions import.* AirflowConfigException"
```

potiuk replied:
Will come back to it some day :).
